File: windows/claude_proxy_bridge.py
#!/usr/bin/env python3
# claude_proxy_bridge.py - Multi-tenant Proxy Bridge with SOCKS5 upstream, Bypass & Stats Logging
# Cross-platform (Windows & Linux) version for Claude Code Proxy Toolkit

import asyncio
import socket
import struct
import urllib.parse
import ipaddress
import os
import sys
import time
import json
from datetime import datetime, timezone

HOME = os.environ.get("USERPROFILE", "") or os.environ.get("HOME", "")
if not HOME:
    HOME = os.path.expanduser("~")

CONFIG_DIR = os.path.join(HOME, ".config", "claude-proxy")
PROGRAMDATA_DIR = os.environ.get("ProgramData", "C:\\ProgramData")
GLOBAL_CONFIG_DIR = os.path.join(PROGRAMDATA_DIR, "claude-proxy")
CONFIG_CANDIDATES = [
    os.path.join(CONFIG_DIR, "proxy.env"),
    os.path.join(GLOBAL_CONFIG_DIR, "proxy.env"),
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "proxy.env")
]
BYPASS_FILE = os.path.join(CONFIG_DIR, "bypass.conf")
LOCAL_BYPASS_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "bypass.conf")
GLOBAL_BYPASS_FILE = os.path.join(GLOBAL_CONFIG_DIR, "bypass.conf")
LOG_FILE = os.path.join(CONFIG_DIR, "traffic.jsonl")

PROXY_PORT = 19000
SOCKS5_URL = ""
PROXY_MODE = "anthropic_only"

def load_env():
    global PROXY_PORT, SOCKS5_URL, PROXY_MODE
    for conf_path in CONFIG_CANDIDATES:
        if os.path.isfile(conf_path):
            try:
                with open(conf_path, "r", encoding="utf-8") as f:
                    for line in f:
                        line = line.strip()
                        if not line or line.startswith("#") or "=" not in line:
                            continue
                        k, v = line.split("=", 1)
                        k = k.strip()
                        v = v.strip().strip("\"'")
                        if k == "PROXY_PORT":
                            PROXY_PORT = int(v)
                        elif k == "SOCKS5_URL":
                            SOCKS5_URL = v
                        elif k == "PROXY_MODE":
                            PROXY_MODE = v
                if SOCKS5_URL:
                    break
            except Exception as e:
                sys.stderr.write(f"[bridge] Warning reading {conf_path}: {e}\n")

load_env()

if not SOCKS5_URL and len(sys.argv) > 1 and not sys.argv[1].startswith("--"):
    SOCKS5_URL = sys.argv[1]

if len(sys.argv) > 2:
    try:
        PROXY_PORT = int(sys.argv[2])
    except ValueError:
        pass

if not SOCKS5_URL:
    sys.stderr.write("[bridge] Error: SOCKS5_URL is not set.\n")
    sys.stderr.write("Usage: python claude_proxy_bridge.py \"socks5://user:pass@host:port\" [port]\n")
    sys.exit(1)

parsed_socks = urllib.parse.urlparse(SOCKS5_URL)
SOCKS5_HOST = parsed_socks.hostname or "127.0.0.1"
SOCKS5_PORT = parsed_socks.port or 1080
SOCKS5_USER = urllib.parse.unquote(parsed_socks.username or "")
SOCKS5_PASS = urllib.parse.unquote(parsed_socks.password or "")

def load_bypass_rules():
    rules = set()
    for p in [LOCAL_BYPASS_FILE, BYPASS_FILE, GLOBAL_BYPASS_FILE]:
        if os.path.isfile(p):
            try:
                with open(p, "r", encoding="utf-8") as f:
                    for line in f:
                        line = line.strip()
                        if line and not line.startswith("#"):
                            rules.add(line.lower())
            except Exception:
                pass
    if not rules:
        rules = {
            "localhost", "127.0.0.1", "::1", "10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16",
            "github.com", "*.github.com", "raw.githubusercontent.com", "objects.githubusercontent.com",
            "avatars.githubusercontent.com", "user-images.githubusercontent.com",
            "gitlab.com", "*.gitlab.com", "bitbucket.org", "*.bitbucket.org",
            "npmjs.org", "*.npmjs.org", "yarnpkg.com", "*.yarnpkg.com",
            "pypi.org", "*.pypi.org", "pythonhosted.org", "*.pythonhosted.org",
            "crates.io", "*.crates.io", "pkg.go.dev", "proxy.golang.org",
            "docker.io", "*.docker.io", "docker.com", "*.docker.com",
            "*.ru", "*.рф", "*.su"
        }
    return rules

BYPASS_RULES = load_bypass_rules()

def is_bypassed(target_host):
    target = target_host.lower().split(":")[0]
    try:
        ip = ipaddress.ip_address(target)
        for rule in BYPASS_RULES:
            if "/" in rule:
                try:
                    if ip in ipaddress.ip_network(rule, strict=False):
                        return True
                except ValueError:
                    pass
            elif rule == target:
                return True
        return False
    except ValueError:
        pass

    for rule in BYPASS_RULES:
        if rule.startswith("*."):
            suffix = rule[1:]
            if target.endswith(suffix) or target == rule[2:]:
                return True
        elif rule.startswith("."):
            if target.endswith(rule) or target == rule[1:]:
                return True
        elif rule == target:
            return True
        elif target.endswith("." + rule):
            return True
            
    return False

def get_process_info(client_port):
    """Detects PID and process name associated with local client_port."""
    if os.name == "nt":
        # Windows: default to claude or use netstat / Get-NetTCPConnection
        return 0, "claude"
    try:
        port_hex = f"{client_port:04X}"
        inode = None
        for proto in ["tcp", "tcp6"]:
            path = f"/proc/net/{proto}"
            if not os.path.exists(path): continue
            with open(path, "r") as f:
                for line in f:
                    parts = line.strip().split()
                    if len(parts) >= 10 and parts[1].endswith(":" + port_hex):
                        inode = parts[9]
                        break
            if inode and inode != "0": break
            
        if not inode or inode == "0":
            return 0, "unknown"
            
        target = f"socket:[{inode}]"
        for pid_dir in os.listdir("/proc"):
            if not pid_dir.isdigit(): continue
            fd_dir = f"/proc/{pid_dir}/fd"
            if not os.path.exists(fd_dir): continue
            try:
                for fd in os.listdir(fd_dir):
                    if os.readlink(f"{fd_dir}/{fd}") == target:
                        pid = int(pid_dir)
                        comm_file = f"/proc/{pid}/comm"
                        cmd_file = f"/proc/{pid}/cmdline"
                        app_name = "unknown"
                        if os.path.exists(comm_file):
                            with open(comm_file, "r") as cf:
                                app_name = cf.read().strip()
                        if os.path.exists(cmd_file):
                            with open(cmd_file, "rb") as cf:
                                cmd_str = cf.read().replace(b"\x00", b" ").decode(errors="ignore").lower()
                            if "claude" in cmd_str:
                                app_name = "claude"
                            elif "copilot" in cmd_str or "codex" in cmd_str:
                                app_name = "copilot"
                            elif "antigravity" in cmd_str or "agy" in cmd_str:
                                app_name = "antigravity"
                        return pid, app_name
            except Exception:
                continue
    except Exception:
        pass
    return 0, "unknown"

def log_traffic(client_ip, client_port, app_name, method, host, port, route, status, bytes_up, bytes_down, latency_ms):
    os.makedirs(CONFIG_DIR, exist_ok=True)
    entry = {
        "ts": datetime.now(timezone.utc).isoformat(),
        "client_ip": client_ip,
        "client_port": client_port,
        "app": app_name,
        "method": method,
        "host": host,
        "port": port,
        "route": route,
        "status": status,
        "bytes_up": bytes_up,
        "bytes_down": bytes_down,
        "latency_ms": round(latency_ms, 2)
    }
    try:
        with open(LOG_FILE, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry, ensure_ascii=False) + "\n")
    except Exception:
        pass

async def socks5_handshake(reader, writer, target_host, target_port):
    if SOCKS5_USER and SOCKS5_PASS:
        writer.write(b"\x05\x02\x00\x02")
    else:
        writer.write(b"\x05\x01\x00")
    await writer.drain()

    resp = await reader.readexactly(2)
    if resp[0] != 5:
        raise Exception(f"Invalid SOCKS5 version: {resp[0]}")
        
    auth_method = resp[1]
    if auth_method == 0x02:
        u_bytes = SOCKS5_USER.encode()
        p_bytes = SOCKS5_PASS.encode()
        auth_pkt = bytes([1, len(u_bytes)]) + u_bytes + bytes([len(p_bytes)]) + p_bytes
        writer.write(auth_pkt)
        await writer.drain()
        auth_resp = await reader.readexactly(2)
        if auth_resp[1] != 0:
            raise Exception("SOCKS5 authentication failed")
    elif auth_method != 0x00:
        raise Exception(f"SOCKS5 auth method {auth_method} not supported")

    host_bytes = target_host.encode("utf-8")
    req = bytes([5, 1, 0, 3, len(host_bytes)]) + host_bytes + struct.pack("!H", target_port)
    writer.write(req)
    await writer.drain()

    conn_resp = await reader.readexactly(4)
    if conn_resp[1] != 0:
        raise Exception(f"SOCKS5 connect error: {conn_resp[1]}")

    atyp = conn_resp[3]
    if atyp == 1:
        await reader.readexactly(4 + 2)
    elif atyp == 3:
        dlen = (await reader.readexactly(1))[0]
        await reader.readexactly(dlen + 2)
    elif atyp == 4:
        await reader.readexactly(16 + 2)

async def pipe_data(src_reader, dst_writer, counter):
    try:
        while True:
            data = await src_reader.read(65536)
            if not data:
                break
            counter[0] += len(data)
            dst_writer.write(data)
            await dst_writer.drain()
    except Exception:
        pass
    finally:
        try:
            dst_writer.close()
        except Exception:
            pass

async def handle_client(client_reader, client_writer):
    client_addr = client_writer.get_extra_info("peername")
    client_ip = client_addr[0] if client_addr else "127.0.0.1"
    client_port = client_addr[1] if client_addr else 0

    _, app_name = get_process_info(client_port)

    start_time = time.time()
    bytes_up = [0]
    bytes_down = [0]
    status_code = 200
    route = "UNKNOWN"
    target_host = "unknown"
    target_port = 443
    method = "UNKNOWN"

    try:
        line = await asyncio.wait_for(client_reader.readline(), timeout=10.0)
        if not line:
            return
        parts = line.decode("utf-8", errors="ignore").strip().split()
        if len(parts) < 2:
            return
        method = parts[0].upper()
        raw_target = parts[1]

        headers = []
        while True:
            h_line = await asyncio.wait_for(client_reader.readline(), timeout=10.0)
            if not h_line or h_line == b"\r\n" or h_line == b"\n":
                break
        if raw_target in ["/proxy.pac", "/pac"]:
            pac_body = (
                "function FindProxyForURL(url, host) {\n"
                "    if (shExpMatch(host, '*.anthropic.com') ||\n"
                "        shExpMatch(host, 'anthropic.com') ||\n"
                "        shExpMatch(host, '*.claude.ai') ||\n"
                "        shExpMatch(host, 'claude.ai') ||\n"
                "        shExpMatch(host, '*.usefathom.com') ||\n"
                "        shExpMatch(host, '*.statsig.com') ||\n"
                "        shExpMatch(host, '*.growthbook.io') ||\n"
                "        shExpMatch(host, '*.datadoghq.com')) {\n"
                f"        return 'PROXY 127.0.0.1:{PROXY_PORT}; DIRECT';\n"
                "    }\n"
                "    return 'DIRECT';\n"
                "}\n"
            ).encode("utf-8")
            resp = (
                b"HTTP/1.1 200 OK\r\n"
                b"Content-Type: application/x-ns-proxy-autoconfig\r\n"
                b"Cache-Control: no-cache\r\n"
                f"Content-Length: {len(pac_body)}\r\n\r\n".encode()
                + pac_body
            )
            client_writer.write(resp)
            await client_writer.drain()
            return

        if method == "CONNECT":
            if ":" in raw_target:
                target_host, port_str = raw_target.split(":", 1)
                target_port = int(port_str)
            else:
                target_host = raw_target
                target_port = 443

            if is_bypassed(target_host):
                route = "DIRECT"
                try:
                    remote_reader, remote_writer = await asyncio.open_connection(target_host, target_port)
                except Exception as e:
                    status_code = 502
                    client_writer.write(b"HTTP/1.1 502 Bad Gateway\r\n\r\n")
                    await client_writer.drain()
                    return
            else:
                route = "SOCKS5"
                try:
                    remote_reader, remote_writer = await asyncio.open_connection(SOCKS5_HOST, SOCKS5_PORT)
                    await socks5_handshake(remote_reader, remote_writer, target_host, target_port)
                except Exception as e:
                    status_code = 502
                    client_writer.write(b"HTTP/1.1 502 Bad Gateway\r\n\r\n")
                    await client_writer.drain()
                    return

            client_writer.write(b"HTTP/1.1 200 Connection Established\r\n\r\n")
            await client_writer.drain()

            t1 = asyncio.create_task(pipe_data(client_reader, remote_writer, bytes_up))
            t2 = asyncio.create_task(pipe_data(remote_reader, client_writer, bytes_down))
            await asyncio.gather(t1, t2)

        else:
            # Regular HTTP GET/POST
            target_host = "unknown"
            for h in headers:
                if h.lower().startswith(b"host:"):
                    target_host = h.split(b":", 1)[1].decode(errors="ignore").strip().split(":")[0]
                    break
            target_port = 80
            if is_bypassed(target_host):
                route = "DIRECT"
                remote_reader, remote_writer = await asyncio.open_connection(target_host, target_port)
            else:
                route = "SOCKS5"
                remote_reader, remote_writer = await asyncio.open_connection(SOCKS5_HOST, SOCKS5_PORT)
                await socks5_handshake(remote_reader, remote_writer, target_host, target_port)

            remote_writer.write(line)
            for h in headers:
                remote_writer.write(h)
            remote_writer.write(b"\r\n")
            await remote_writer.drain()

            t1 = asyncio.create_task(pipe_data(client_reader, remote_writer, bytes_up))
            t2 = asyncio.create_task(pipe_data(remote_reader, client_writer, bytes_down))
            await asyncio.gather(t1, t2)

    except Exception:
        status_code = 502
    finally:
        try:
            client_writer.close()
        except Exception:
            pass
        latency_ms = (time.time() - start_time) * 1000.0
        log_traffic(client_ip, client_port, app_name, method, target_host, target_port, route, status_code, bytes_up[0], bytes_down[0], latency_ms)

async def main():
    server = await asyncio.start_server(handle_client, "127.0.0.1", PROXY_PORT)
    print(f"[bridge] Python Proxy Bridge running on 127.0.0.1:{PROXY_PORT}")
    print(f"[bridge] Upstream SOCKS5: socks5://{SOCKS5_HOST}:{SOCKS5_PORT}")
    print(f"[bridge] Active bypass rules: {len(BYPASS_RULES)} entries")
    print(f"[bridge] Audit log: {LOG_FILE}")
    async with server:
        await server.serve_forever()

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass

File: windows/claude_proxy_bridge.js
#!/usr/bin/env node
/**
 * claude_proxy_bridge.js - Cross-platform Proxy Bridge for Claude Code (Windows / Node.js)
 * SOCKS5 upstream, Destination Bypass, Traffic Audit Logging & Fail-Closed security.
 * Part of Claude Code Proxy Toolkit
 *
 * Usage:
 *   node claude_proxy_bridge.js [SOCKS5_URL] [PORT]
 */

const http = require('http');
const net = require('net');
const fs = require('fs');
const path = require('path');
const os = require('os');
const { URL } = require('url');

// Environment & paths
const HOME = os.homedir();
const CONFIG_DIR = path.join(HOME, '.config', 'claude-proxy');
const PROGRAMDATA_DIR = process.env.ProgramData ? path.join(process.env.ProgramData, 'claude-proxy') : 'C:\\ProgramData\\claude-proxy';
const CONFIG_CANDIDATES = [
    path.join(CONFIG_DIR, 'proxy.env'),
    path.join(PROGRAMDATA_DIR, 'proxy.env'),
    path.join(__dirname, 'proxy.env'),
    path.join(__dirname, '..', 'proxy.env')
];
const BYPASS_FILE = path.join(CONFIG_DIR, 'bypass.conf');
const LOCAL_BYPASS_FILE = path.join(__dirname, 'bypass.conf');
const PROGRAMDATA_BYPASS_FILE = path.join(PROGRAMDATA_DIR, 'bypass.conf');
const LOG_FILE = path.join(CONFIG_DIR, 'traffic.jsonl');

let PROXY_PORT = 19000;
let SOCKS5_URL = '';
let PROXY_MODE = 'anthropic_only';

// 1. Load config from proxy.env
function loadConfig() {
    for (const confPath of CONFIG_CANDIDATES) {
        if (fs.existsSync(confPath)) {
            try {
                const content = fs.readFileSync(confPath, 'utf8');
                for (let line of content.split('\n')) {
                    line = line.trim();
                    if (!line || line.startsWith('#') || !line.includes('=')) continue;
                    const eq = line.indexOf('=');
                    const k = line.substring(0, eq).trim();
                    const v = line.substring(eq + 1).trim().replace(/^["']|["']$/g, '');
                    if (k === 'PROXY_PORT') PROXY_PORT = parseInt(v, 10) || 19000;
                    else if (k === 'SOCKS5_URL') SOCKS5_URL = v;
                    else if (k === 'PROXY_MODE') PROXY_MODE = v;
                }
                if (SOCKS5_URL) break; // Successfully loaded SOCKS5_URL
            } catch (err) {
                console.error('[bridge] Warning reading config:', confPath, err.message);
            }
        }
    }
}

loadConfig();

// Command line overrides
if (process.argv[2] && !process.argv[2].startsWith('--')) {
    SOCKS5_URL = process.argv[2];
}
if (process.argv[3] && !isNaN(parseInt(process.argv[3], 10))) {
    PROXY_PORT = parseInt(process.argv[3], 10);
}

if (!SOCKS5_URL) {
    console.error('[bridge] Error: SOCKS5_URL is not set in config or command line arguments.');
    console.error('Usage: node claude_proxy_bridge.js "socks5://user:pass@host:port" [port]');
    process.exit(1);
}

// Parse SOCKS5 URL
let parsedSocks;
try {
    let urlStr = SOCKS5_URL;
    if (!urlStr.includes('://')) urlStr = 'socks5://' + urlStr;
    parsedSocks = new URL(urlStr);
} catch (e) {
    console.error('[bridge] Invalid SOCKS5 URL:', SOCKS5_URL);
    process.exit(1);
}

const SOCKS_HOST = parsedSocks.hostname;
const SOCKS_PORT = parseInt(parsedSocks.port, 10) || 1080;
const SOCKS_USER = decodeURIComponent(parsedSocks.username || '');
const SOCKS_PASS = decodeURIComponent(parsedSocks.password || '');

// 2. Load Bypass rules
let bypassRules = new Set();

function loadBypassRules() {
    const rules = new Set();
    const filesToRead = [LOCAL_BYPASS_FILE, BYPASS_FILE, PROGRAMDATA_BYPASS_FILE];
    for (const f of filesToRead) {
        if (fs.existsSync(f)) {
            try {
                const lines = fs.readFileSync(f, 'utf8').split('\n');
                for (let line of lines) {
                    line = line.trim().toLowerCase();
                    if (line && !line.startsWith('#')) {
                        rules.add(line);
                    }
                }
            } catch {}
        }
    }
    if (rules.size === 0) {
        [
            'localhost', '127.0.0.1', '::1', '10.0.0.0/8', '172.16.0.0/12', '192.168.0.0/16',
            'github.com', '*.github.com', 'raw.githubusercontent.com', 'objects.githubusercontent.com',
            'gitlab.com', '*.gitlab.com', 'bitbucket.org', '*.bitbucket.org',
            'npmjs.org', '*.npmjs.org', 'registry.npmjs.org', 'yarnpkg.com', '*.yarnpkg.com',
            'pypi.org', '*.pypi.org', 'pythonhosted.org', '*.pythonhosted.org',
            'crates.io', '*.crates.io', 'pkg.go.dev', 'proxy.golang.org',
            'docker.io', '*.docker.io', 'docker.com', '*.docker.com',
            '*.ru', '*.рф', '*.su'
        ].forEach(r => rules.add(r.toLowerCase()));
    }
    bypassRules = rules;
}

loadBypassRules();
setInterval(loadBypassRules, 15000); // Reload bypass rules every 15s

function isBypassed(host) {
    if (!host) return false;
    const target = host.toLowerCase().split(':')[0];

    for (const rule of bypassRules) {
        if (rule === target) return true;
        if (rule.startsWith('*.')) {
            const suffix = rule.substring(1); // e.g. .github.com
            if (target.endsWith(suffix) || target === rule.substring(2)) return true;
        } else if (rule.startsWith('.')) {
            if (target.endsWith(rule) || target === rule.substring(1)) return true;
        } else if (target.endsWith('.' + rule)) {
            return true;
        }
    }
    return false;
}

// 3. Traffic Logging
if (!fs.existsSync(CONFIG_DIR)) {
    try { fs.mkdirSync(CONFIG_DIR, { recursive: true }); } catch {}
}

function logTraffic(entry) {
    const line = JSON.stringify({
        ts: new Date().toISOString(),
        client_ip: entry.clientIp || '127.0.0.1',
        client_port: entry.clientPort || 0,
        app: entry.app || 'claude',
        method: entry.method || 'CONNECT',
        host: entry.host,
        port: entry.port,
        route: entry.route,
        status: entry.status,
        bytes_up: entry.bytesUp || 0,
        bytes_down: entry.bytesDown || 0,
        latency_ms: entry.latencyMs || 0
    }) + '\n';

    fs.appendFile(LOG_FILE, line, (err) => {
        if (err && err.code !== 'ENOENT') {
            // Ignore logging errors silently
        }
    });
}

// 4. SOCKS5 Connect Function
function connectSocks5(targetHost, targetPort, callback) {
    const socket = net.connect(SOCKS_PORT, SOCKS_HOST);
    let stage = 0; // 0: init, 1: auth, 2: connect
    let handshakeFinished = false;

    socket.setTimeout(15000, () => {
        socket.destroy(new Error('SOCKS5 connection timeout'));
    });

    socket.on('connect', () => {
        // Step 1: Greeting
        if (SOCKS_USER && SOCKS_PASS) {
            socket.write(Buffer.from([0x05, 0x02, 0x00, 0x02])); // Support NO_AUTH and USER/PASS
        } else {
            socket.write(Buffer.from([0x05, 0x01, 0x00])); // Support NO_AUTH
        }
    });

    socket.on('data', (chunk) => {
        if (handshakeFinished) return;

        if (stage === 0) {
            // Greeting response
            if (chunk.length < 2 || chunk[0] !== 0x05) {
                socket.destroy(new Error('Invalid SOCKS5 version in greeting response'));
                return;
            }
            const authMethod = chunk[1];
            if (authMethod === 0x00) {
                // No auth required, proceed to connect
                sendConnectRequest();
            } else if (authMethod === 0x02) {
                // Username/Password authentication
                stage = 1;
                const uBuf = Buffer.from(SOCKS_USER);
                const pBuf = Buffer.from(SOCKS_PASS);
                const authPacket = Buffer.concat([
                    Buffer.from([0x01, uBuf.length]),
                    uBuf,
                    Buffer.from([pBuf.length]),
                    pBuf
                ]);
                socket.write(authPacket);
            } else {
                socket.destroy(new Error(`Unsupported SOCKS5 auth method: ${authMethod}`));
            }
        } else if (stage === 1) {
            // Auth response
            if (chunk.length < 2 || chunk[1] !== 0x00) {
                socket.destroy(new Error('SOCKS5 authentication failed'));
                return;
            }
            sendConnectRequest();
        } else if (stage === 2) {
            // Connect response
            if (chunk.length < 4 || chunk[1] !== 0x00) {
                const repCode = chunk.length >= 2 ? chunk[1] : -1;
                socket.destroy(new Error(`SOCKS5 connect rejected, rep=${repCode}`));
                return;
            }
            // SOCKS5 tunnel established!
            handshakeFinished = true;
            socket.setTimeout(0);
            callback(null, socket);
        }
    });

    function sendConnectRequest() {
        stage = 2;
        const hostBuf = Buffer.from(targetHost);
        const portBuf = Buffer.alloc(2);
        portBuf.writeUInt16BE(targetPort, 0);

        // Address type 0x03 (Domain Name) -> DNS is resolved by the remote SOCKS5 proxy!
        const reqPacket = Buffer.concat([
            Buffer.from([0x05, 0x01, 0x00, 0x03, hostBuf.length]),
            hostBuf,
            portBuf
        ]);
        socket.write(reqPacket);
    }

    socket.on('error', (err) => {
        if (!handshakeFinished) {
            callback(err);
        }
    });
}

// 5. Create HTTP Server (handles CONNECT tunneling, regular HTTP and PAC serving)
const server = http.createServer((req, res) => {
    // Serve PAC (Proxy Auto-Configuration) file for Windows & Claude Desktop GUI
    if (req.url === '/proxy.pac' || req.url === '/pac') {
        res.writeHead(200, {
            'Content-Type': 'application/x-ns-proxy-autoconfig',
            'Cache-Control': 'no-cache'
        });
        const pac = `function FindProxyForURL(url, host) {
    // Only Anthropic and Claude domains route through local SOCKS5 bridge
    if (shExpMatch(host, "*.anthropic.com") ||
        shExpMatch(host, "anthropic.com") ||
        shExpMatch(host, "*.claude.ai") ||
        shExpMatch(host, "claude.ai") ||
        shExpMatch(host, "*.usefathom.com") ||
        shExpMatch(host, "*.statsig.com") ||
        shExpMatch(host, "*.growthbook.io") ||
        shExpMatch(host, "*.datadoghq.com")) {
        return "PROXY 127.0.0.1:${PROXY_PORT}; DIRECT";
    }
    // Everything else connects directly (no proxy overhead or interference)
    return "DIRECT";
}
`;
        res.end(pac);
        return;
    }

    // Plain HTTP proxy request
    const startTime = Date.now();
    let host = req.headers.host || '';
    let port = 80;
    if (host.includes(':')) {
        const parts = host.split(':');
        host = parts[0];
        port = parseInt(parts[1], 10) || 80;
    }

    const bypassed = isBypassed(host);
    const route = bypassed ? 'DIRECT' : 'SOCKS5';

    if (bypassed) {
        const proxyReq = http.request({
            host: host,
            port: port,
            path: req.url,
            method: req.method,
            headers: req.headers
        }, (proxyRes) => {
            res.writeHead(proxyRes.statusCode, proxyRes.headers);
            proxyRes.pipe(res);
            logTraffic({
                clientIp: req.socket.remoteAddress,
                clientPort: req.socket.remotePort,
                app: 'claude',
                method: req.method,
                host: host,
                port: port,
                route: route,
                status: proxyRes.statusCode,
                latencyMs: Date.now() - startTime
            });
        });

        proxyReq.on('error', (err) => {
            res.writeHead(502);
            res.end('Bad Gateway');
            logTraffic({
                clientIp: req.socket.remoteAddress,
                clientPort: req.socket.remotePort,
                app: 'claude',
                method: req.method,
                host: host,
                port: port,
                route: route,
                status: 502,
                latencyMs: Date.now() - startTime
            });
        });

        req.pipe(proxyReq);
    } else {
        // Forward HTTP via SOCKS5
        connectSocks5(host, port, (err, socksSocket) => {
            if (err) {
                res.writeHead(502);
                res.end('Bad Gateway: ' + err.message);
                logTraffic({
                    clientIp: req.socket.remoteAddress,
                    clientPort: req.socket.remotePort,
                    app: 'claude',
                    method: req.method,
                    host: host,
                    port: port,
                    route: route,
                    status: 502,
                    latencyMs: Date.now() - startTime
                });
                return;
            }

            // Write raw HTTP request to SOCKS5 socket
            socksSocket.write(`${req.method} ${req.url} HTTP/1.1\r\n`);
            for (const [key, val] of Object.entries(req.headers)) {
                socksSocket.write(`${key}: ${val}\r\n`);
            }
            socksSocket.write('\r\n');

            req.pipe(socksSocket);
            socksSocket.pipe(res.socket);
        });
    }
});

// Handle HTTPS CONNECT tunneling
server.on('connect', (req, clientSocket, head) => {
    const startTime = Date.now();
    const [targetHost, targetPortStr] = req.url.split(':');
    const targetPort = parseInt(targetPortStr, 10) || 443;

    const bypassed = isBypassed(targetHost);
    const route = bypassed ? 'DIRECT' : 'SOCKS5';

    let bytesUp = 0;
    let bytesDown = 0;
    let finished = false;

    function finishLog(status) {
        if (finished) return;
        finished = true;
        logTraffic({
            clientIp: clientSocket.remoteAddress,
            clientPort: clientSocket.remotePort,
            app: 'claude',
            method: 'CONNECT',
            host: targetHost,
            port: targetPort,
            route: route,
            status: status,
            bytesUp: bytesUp,
            bytesDown: bytesDown,
            latencyMs: Date.now() - startTime
        });
    }

    clientSocket.on('data', (d) => { bytesUp += d.length; });

    if (bypassed) {
        // Direct connection
        const upstreamSocket = net.connect(targetPort, targetHost, () => {
            clientSocket.write('HTTP/1.1 200 Connection Established\r\n\r\n');
            if (head && head.length > 0) upstreamSocket.write(head);
            upstreamSocket.pipe(clientSocket);
            clientSocket.pipe(upstreamSocket);
            finishLog(200);
        });

        upstreamSocket.on('data', (d) => { bytesDown += d.length; });

        upstreamSocket.on('error', (err) => {
            if (!clientSocket.destroyed) {
                clientSocket.write('HTTP/1.1 502 Bad Gateway\r\n\r\n');
                clientSocket.end();
            }
            finishLog(502);
        });

        clientSocket.on('error', () => { upstreamSocket.destroy(); });
    } else {
        // Connect via SOCKS5
        connectSocks5(targetHost, targetPort, (err, socksSocket) => {
            if (err) {
                if (!clientSocket.destroyed) {
                    clientSocket.write('HTTP/1.1 502 Bad Gateway\r\n\r\n');
                    clientSocket.end();
                }
                finishLog(502);
                return;
            }

            clientSocket.write('HTTP/1.1 200 Connection Established\r\n\r\n');
            if (head && head.length > 0) socksSocket.write(head);

            socksSocket.on('data', (d) => { bytesDown += d.length; });

            socksSocket.pipe(clientSocket);
            clientSocket.pipe(socksSocket);

            finishLog(200);

            socksSocket.on('error', () => { clientSocket.destroy(); });
            clientSocket.on('error', () => { socksSocket.destroy(); });
        });
    }
});

server.listen(PROXY_PORT, '127.0.0.1', () => {
    console.log(`[bridge] Claude Code Proxy Bridge running on http://127.0.0.1:${PROXY_PORT}`);
    console.log(`[bridge] Upstream SOCKS5: socks5://${SOCKS_HOST}:${SOCKS_PORT} (User: ${SOCKS_USER ? SOCKS_USER : 'none'})`);
    console.log(`[bridge] Active bypass rules: ${bypassRules.size} entries`);
    console.log(`[bridge] Audit log: ${LOG_FILE}`);
});

server.on('error', (err) => {
    console.error(`[bridge] Fatal server error:`, err.message);
    process.exit(1);
});
